### PR TITLE
Release/4.2.3.0.0

### PR DESCRIPTION
**Pull request by ricealexanderb**
There's no Rewarded ad to test this with yet but Banner and Interstitial work.
I'm traveling the next four days but will address any feedback as soon as I return on Tuesday, and hopefully submit a release candidate soon after that.

### CODE REVIEW

File: Source/BidMachineAdapter.swift
Comment (daniel-barros) [inline Source/BidMachineAdapter.swift:88]: Just to confirm, calls to `BidMachineSdk.shared.regulationInfo.populate` have no effect after initialization?
If they did we could call it here too

File: Source/BidMachineAdapter.swift
Comment (daniel-barros) [inline Source/BidMachineAdapter.swift]: Do we need to add support for adaptive banner here?
@kendallrogers

File: Source/BidMachineAdapter.swift
Comment (daniel-barros) [inline Source/BidMachineAdapter.swift:138]: Have we tested that loading multiple ads with the same placement works with this partner?
If not we may need to add a safeguard here similar to what we have for other partners.

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]:
```suggestion
            log(.loadFailed(error))
            completion(.failure(error))
```

Comment (daniel-barros) [reply Source/BidMachineAdapterBannerAd.swift]: Mediation SDK takes care of wrapping the partner error if it's not a ChartboostMediationError

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift:60]: Can you capture the viewController weakly so we don't keep it alive longer than the publisher does?

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]:
```suggestion
            loadCompletion?(.failure(loadError)) ?? log(.loadResultIgnored)
```

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]: Not sure but I think we don't log show-related events for banners in other adapters.
Banners don't have a show process according to Helium

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]: showCompletion should always be nil for banners, since they are never asked to show. Better to remove it to avoid confusion.

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]: Same

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterConfig.swift]: Would using a `nil` instead of a default value work? Could be safer than specifying a string.

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterConfig.swift]: Use `UserDefaults.standard.value(forKey:)` and cast it instead. That way you don't have the issue of default booleans getting returned.

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterConfig.swift]: Can we just make it optional and return `nil` if nothing has been set?

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterConfig.swift]: Same

Comment (daniel-barros) [reply Source/BidMachineAdapterConfig.swift]: And everywhere else

File: Source/BidMachineAdapterInterstitialAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterInterstitialAd.swift]:
```suggestion
            log(.loadFailed(error))
            completion(.failure(error))
```

File: Source/BidMachineAdapterInterstitialAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterInterstitialAd.swift]: We should call the didDismiss method on the delegate here, and log the proper event

File: Source/BidMachineAdapterRewardedAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterRewardedAd.swift]:
```suggestion
            log(.loadFailed(error))
            completion(.failure(error))
```

File: Source/BidMachineAdapterRewardedAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterRewardedAd.swift]: Same as with interstitial, let's report the dismiss

File: Source/BidMachineAdapter.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapter.swift:88]: Their [documentation](https://docs.bidmachine.io/docs/in-house-mediation-1) doesn't say, and their [example code](https://github.com/bidmachine/BidMachine-iOS-Examples/blob/master/Swift/Banner/Banner/AppDelegate.swift) only shows it being used immediately before init. Today I tried to prove that calling `populate` after init doesn't make a difference, but I stopped receiving fill regardless of how test mode was set at init and I also didn't see any log messages in the console even when logging was turned on at init, so I can't prove that calling `populate` here wouldn't work but I also don't have any reason to believe it would.

File: Source/BidMachineAdapter.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapter.swift:138]: I had not tested it until you reminded me, and BidMachine does fail when the same placement is loaded multiple times so [I have added the safeguard](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/commit/b9a0bbd50681c2aaf58f21c0c1e94030b844b3f6).

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: The completion's signature is `(BidMachine.BidMachineBanner?, Error?) -> Void` so `error` is an optional and can't be passed directly to `.loadFailed()`

The current code is the most concise way I see of handling that, but if it's a priority to hand back an unwrapped partner error whenever possible I could change this section to check that `error` is not nil and only make a new error object if necessary.

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: Looks like the only other banner ad that logs a show error is Vungle, and I was the one who put the log statement there. So there is not a strong precedent for doing this.

But when the delegate protocol has both "failed to load" and "failed to show" methods, I think it's better not to hide that extra debugging information they're passing to us. And I don't think there's a contradiction between making "show" a sub-process of loading but reporting that something happened specifically while trying to display the ad.

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: I can't change the `Bool`s in `BidMachineAdapterConfig` to `Bool?` unless I remove `@objc`, but I think the adapter config needs to be accessible from Objective-C doesn't it?

Maybe I should make these properties private and only expose setter methods to publishers. For example: if the publisher has never called `BidMachineAdapterConfig.setGDPRConsent(...)` then the private `BidMachineAdapterConfig.gdprConsent` value checked during adapter init will be nil.

File: Source/BidMachineAdapter.swift
Comment (kendallrogers) [reply Source/BidMachineAdapter.swift]: Yeah, you can just follow the pattern [here](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-adcolony/blob/227d0c94b6eeab41f460f47c9ece1ca3b084e19e/Source/AdColonyAdapter.swift#L156-L161).

File: Source/BidMachineAdapterBannerAd.swift
Comment (kendallrogers) [inline Source/BidMachineAdapterBannerAd.swift]: For adaptive banner support, the size should be calculated like it is [here](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-adcolony/blob/227d0c94b6eeab41f460f47c9ece1ca3b084e19e/Source/AdColonyAdapterBannerAd.swift#L95-L109).

File: Source/BidMachineAdapterBannerAd.swift
Comment (kendallrogers) [inline Source/BidMachineAdapterBannerAd.swift]: For adaptive banner support, we should return the loaded ad size like we do [here](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-adcolony/blob/227d0c94b6eeab41f460f47c9ece1ca3b084e19e/Source/AdColonyAdapterBannerAd.swift#L68-L75). If there's a way in the BidMachine API to get the size of the loaded ad from the view, we can use that, but if not, just using the requested size should be fine.

File: Source/BidMachineAdapter.swift
Comment (daniel-barros) [reply Source/BidMachineAdapter.swift:88]: Sees like there's no downside in adding support for changing consent post-initialization, if it doesn't do anything then it doesn't do anything, and there's the upside that it may work.
If we can't prove it doesn't work I think it's better to add support for it.

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [reply Source/BidMachineAdapterBannerAd.swift]: Ah gotcha. Nvm then.

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [reply Source/BidMachineAdapterBannerAd.swift]: We could still log the call to that method using the `delegateCallIgnored` log. I think this makes it clearer that the delegate call has no effect and is not part of any of the main ad lifecycle processes handled by Helium.
Seems better for consistency with other adapters since as you said there's no strong precedence for doing it the other way. This is a bit nit-picky though.

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [reply Source/BidMachineAdapterConfig.swift]: Ahhh gotcha. Yeah your idea of separating setters and getters sounds good to me.

File: Source/BidMachineAdapter.swift
Comment (daniel-barros) [reply Source/BidMachineAdapter.swift:138]: I recently updated the adapter development guide to include this check as a step. Hopefully that'll help us remember this.

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: Changed in [this commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/commit/d8566bcabd48c1d3ab857ac6e4d96bbc6d2bfc97).

File: Source/BidMachineAdapter.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapter.swift]: Added in [this commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/commit/d8566bcabd48c1d3ab857ac6e4d96bbc6d2bfc97).

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: I think I was wrong about the privacy settings needing to be available from Objective-C. The logging settings on `BidMachineAdapterConfig` need to be accessible to publishers, but the privacy settings are all set though the mediation SDK.
I've changed them all to `public static var` in [this commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/5769e6494e3a384e5ec3e2c32e2f6b1d98b030e4).

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: I made all the privacy related changes in [this commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/5769e6494e3a384e5ec3e2c32e2f6b1d98b030e4) (same one as the conversation above)

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: At first I thought you meant I should remove it and put `delegateCallIgnored` there instead. Re-reading this conversation, I think maybe the important part is that we log that the delegate call has no effect but you wouldn't necessarily object to also leaving the `showFailed` log here?

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: Maybe I should mark them as `internal`, since there's not a good reason for them to be visible in Swift either?

Comment (daniel-barros) [reply Source/BidMachineAdapterConfig.swift]: If we don't need to make them accessible to publishers and the SDK will set the consent then we can just remove all these `BidMachineAdapterConfig` properties?

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [reply Source/BidMachineAdapterBannerAd.swift]: No, I meant remove `showFailed` and replace it by `delegateCallIgnored`. This seems the coherent thing to do with other adapters (although I might very wrong here since I haven't checked them in a while) and makes it clear that this is not affecting any Helium show operation.
It's fine by me if you choose to keep `showFailed` though. But definitely not both. Note that `delegateCallIgnored` will log the name of the delegate method which is redundant info if we are also logging the `showFailed` event.

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: The properties are stored so that they can be applied when the BidMachine adapter is initialized. We don't know whether calling `BidMachineSdk.shared.regulationInfo.populate` after init does anything.

Maybe I should try getting in touch with someone at BidMachine who can answer that question.

Comment (daniel-barros) [reply Source/BidMachineAdapterConfig.swift]: I'm okay with having them stored, what I mean is if we aren't making them available to publishers, then we should make them private and move them into the main adapter class?

Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: I agree they should be private, but since `BidMachineAdapterConfig` will still be managing the BidMachine logging settings I think it makes sense to put all the configuration options in one place.

Comment (daniel-barros) [reply Source/BidMachineAdapterConfig.swift]: To me it seems that would deviate from the current norm of having consent and consent persistence logic inside the main adapter class, and using the Config class only for public config settings.
Will approve the PR anyway if you strongly feel this is the way.

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: Changed in [570f70831c5e1ae8389c7c1eb446111cc00c7ad7](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/570f70831c5e1ae8389c7c1eb446111cc00c7ad7)

File: Source/BidMachineAdapterInterstitialAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterInterstitialAd.swift]: Discussed in [previous thread](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1#discussion_r1358716581).

File: Source/BidMachineAdapterRewardedAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterRewardedAd.swift]: Discussed in [previous thread](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1#discussion_r1358716581).

File: Source/BidMachineAdapterConfig.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterConfig.swift]: Moved in [e2353772729fb12d49b47ac078df16ad30ba9b26](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/e2353772729fb12d49b47ac078df16ad30ba9b26)

File: Source/BidMachineAdapterBannerAd.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapterBannerAd.swift]: Removed in [e9842b13f3b41404873b6aa7bede38ddd9c8b2b5](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/e9842b13f3b41404873b6aa7bede38ddd9c8b2b5)

File: Source/BidMachineAdapter.swift
Comment (ricealexanderb) [reply Source/BidMachineAdapter.swift:88]: Added in [5769e6494e3a384e5ec3e2c32e2f6b1d98b030e4](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/5769e6494e3a384e5ec3e2c32e2f6b1d98b030e4)

**Comment by ricealexanderb** [comment]
After contacting BidMachine, we now know it's safe to set the regulation info after init, so I removed all the storage code that was used to populate those values before init.
[2e025c5f908a16ee99bff08c7732a277487b3a57](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/pull/1/commits/2e025c5f908a16ee99bff08c7732a277487b3a57)

File: Source/BidMachineAdapterConfig.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterConfig.swift]: nit:
```suggestion
```

File: Source/BidMachineAdapterInterstitialAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterInterstitialAd.swift]: nit:
Inside a `catch` statement error should be non-optional, so we could pass the partner error directly.
```suggestion
            log(.loadFailed(error))
            completion(.failure(error))
```

File: Source/BidMachineAdapterRewardedAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterRewardedAd.swift]: nit: same as in interstitial

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]: nit: same as in interstitial

File: Source/BidMachineAdapterBannerAd.swift
Comment (daniel-barros) [inline Source/BidMachineAdapterBannerAd.swift]: Let's remove `log(.showStarted)` if there's no corresponding showSucceeded log anywhere
```suggestion
```

**Review by daniel-barros — DISMISSED** [review]
Thanks for all the changes and for communicating with the partner Alex. The consent logic looks so much simple now.
I left a few nits but nothing major.

**Comment by ricealexanderb** [comment]
I made all the mentioned changes in [this commit](https://github.com/ChartBoost/chartboost-mediation-ios-adapter-bidmachine/commit/9dd98213c1dde6ff0355320aa7a7a60871d2ba18). Building release candidate now, but need re-approval so this branch can be merged later.

**Review by daniel-barros — DISMISSED** [review]

**Pull request closed without merging**
